src/librustc/infer/resolve.rs

```diff
@@ -104,8 +104,45 @@ impl<'a, 'gcx, 'tcx> TypeVisitor<'tcx> for UnresolvedTypeFinder<'a, 'gcx, 'tcx>
                 t.super_visit_with(self)
             }
         } else {
-            // Micro-optimize: no inference types at all Can't have unresolved type
-            // variables, no need to visit the contents.
+            // All type variables in inference types must already be resolved,
+            // - no need to visit the contents
+            false
+        }
+    }
+}
+
+pub struct UnresolvedTypeCollector<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
```

Comment:
So this looks pretty similar to this code here:

https://github.com/rust-lang/rust/blob/07133ac70cca85b2f91aedb76a21ece524bc0cb4/src/librustc/infer/resolve.rs#L80-L112

Is there a reason we had to copy-and-paste it? I'm a bit confused by this addition.

Reply:
I was worried about the performance impact of holding onto an extra vector, but it probably doesn't make a big difference - you're right less code is more important.